src/rnpkeys/rnpkeys.c

```diff
@@ -234,10 +242,34 @@ rnp_cmd(rnp_t *rnp, prog_t *p, char *f)
         }
         return rnp_import_key(rnp, f);
     case GENERATE_KEY:
-        if ((key = f) == NULL) {
-            key = rnp_getvar(rnp, "userid");
+        key = f ? f : rnp_getvar(rnp, "userid");
+        rnp_keygen_desc_t *ctx = &rnp->action.generate_key_ctx;
+        if (findvar(rnp, "expert") > 0) {
+            (void) rnp_generate_key_expert_mode(rnp);
+        } else {
+            ctx->key_alg = PGP_PKA_RSA;
+            ctx->rsa.modulus_bit_len = p->numbits;
         }
-        return rnp_generate_key(rnp, key, p->numbits);
+
+        // Find hash algorithm to use
+        const char *hash_str = rnp_getvar(rnp, "hash");
+        ctx->hash_alg = hash_str ? pgp_str_to_hash_alg(hash_str) : PGP_HASH_UNKNOWN;
+        if (PGP_HASH_UNKNOWN == ctx->hash_alg) {
+            // Use default hash
+            if (ctx->key_alg == PGP_PKA_ECDSA) {
+                ctx->hash_alg = (ctx->ecc.curve == PGP_CURVE_NIST_P_256) ?
+                                  PGP_HASH_SHA256 :
+                                  (ctx->ecc.curve == PGP_CURVE_NIST_P_384) ?
+                                  PGP_HASH_SHA384 :
+                                  /*PGP_CURVE_NIST_P_521*/ PGP_HASH_SHA512;
+            } else {
+                // TODO: Shouldn't it be PGP_DEFAULT_HASH_ALGORITHM ?
```

Comment:
`PGP_DEFAULT_HASH_ALGORITHM` is now specified as SHA256 in code, but in RFC 4880 (and bis01) "SHA-1" is still the only mandatory implementation? Can we confirm this?